src/python/pants/backend/java/target_types.py

```diff
@@ -142,6 +144,40 @@ async def generate_targets_from_java_sources(
     )
 
 
+# Things for JARs
+#
+
+
+class JvmMainClassName(StringField):
+    alias = "main"
+    required = True
+    help = (
+        "`.`-separated name of the JVM class containing the `main()` method to be called when "
+        "executing this JAR."
+    )
+
+
+class JvmRootClassAddress(StringField):
```

Comment:
Could the address for the main class just be provided via one of the dependencies specified via `dependencies`?

Reply:
Only if we can reliably calculate the Java name for the class. Prior art from most other build tools suggests that this should just be provided separately.

Reply:
> Only if we can reliably calculate the Java name for the class.

You can: there is a determinstic mapping.

But I don't think that that is relevant here: all you are using this for is kicking off a transitive walk on the dependencies. And rather than doing that using an explicit address, you can do it using the target's dependencies. Take a look at the `junit` rule I linked above: it should be similar in many ways, since it is also building a runnable classpath (aside from needing an explicit main class obviously).

Reply:
There's a deterministic mapping for Java files, but it's not immediately clear to me that this applies for other languages.

Reply:
To confirm: the `root_address` is not necessary, and should be removed. The `FieldSet`'s own `Address` is the root, and the main class just needs to exist anywhere in the transitive graph from the root (but it isn't necessary to validate that that is the case).

Reply:
I'll have to follow up out of band here, as it's not completely obvious what I should be doing to change this.